### ISSUE TITLE
Automated testing for Chrome

### ISSUE THREAD

**Issue by jsha** [EFForg/https-everywhere#2230]
I'd very much like to add automated Chrome testing, similar to our Firefox automated test: Start up a Chrome profile with the extension loaded, open a sample URL, and verify that it got rewritten to HTTPS. Some of the groundwork is in run-chromium.sh, but I'm not sure how to proceed with the rest of the test: in particular how to detect once the page is fully loaded and shut down. If you have some ideas, please contribute!

**Comment by semenko** [comment]
Might be messy, though looks like some good info in:

https://github.com/travis-ci/travis-ci/issues/272

**Comment by TheNavigat** [comment]
I've already done something like that for a Chrome extension I've created using Selenium and ChromeDriver. The first thing you'll need to do is to install ChromeDriver itself. If you're on Linux, you'll need to build it manually. I have an executable file I can send you if you'd like. Afterwards, install Selenium. And of course you need to have Google Chrome installed.

Selenium scripts can be coded using either Java or Python, I use Python. Here's an example of loading a Chrome browser with an extension using Python.

```python
import time
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.common.by import By

chromeOps = webdriver.ChromeOptions()
chromeOps.add_extension('Source.crx')

driver = webdriver.Chrome('/usr/bin/chromedriver', chrome_options = chromeOps)  # Optional argument, if not specified will search path.
driver.get('https://google.com')

wait = WebDriverWait(driver, 20)
wait.until(expected_conditions.visibility_of_element_located((By.ID, 'hplogo')))

#Page Loaded
```

I'm willing to contribute but I'm new here, I don't know what exactly do we need to test.

**Comment by reedy** [comment]
I guess the intention would be to do the testing on travis

**Comment by TheNavigat** [comment]
I don't think Travis supports ChromeDriver

**Comment by jsha** [comment]
@reedy is right, the ideal is run automated tests on Travis CI, but according to https://github.com/travis-ci/travis-ci/issues/272, Chrome won't run on Travis. Even so, I think it's worthwhile to have an automated test that developers can run on their workstations, so I greatly appreciate the offer, @TheNavigat!

The most important test I'd like to do is to start up Chrome, navigate to an example rewritten site, e.g. http://libssh.org/robots.txt, wait for the document to load, then verify the document's URL is https://libssh.org/robots.txt. Once we've got that running we can look at more tests, for instance opening the icon menu, ensuring the correct list of applicable rules is there, and enabling / disabling rules.

**Comment by TheNavigat** [comment]
@jsha That's the exact ticket I was looking at as well when I mentioned that Travis won't work.

And I'll get that done in a while.

**Comment by TheNavigat** [comment]
```python
import time
from selenium import webdriver
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.support.wait import WebDriverWait
from selenium.webdriver.support import expected_conditions
from selenium.webdriver.common.by import By

chromeOps = webdriver.ChromeOptions()
chromeOps.add_extension('Source.crx')

driver = webdriver.Chrome('/usr/bin/chromedriver', chrome_options = chromeOps)  # Optional argument, if not specified will search path.
driver.get('http://libssh.org/robots.txt')

#Page Loaded

if driver.current_url.startswith('https'):
	print "HTTP to HTTPS redirection successful"
elif(driver.current_url.startswith('http')):
	print "HTTP to HTTPS redirection failed"

time.sleep(5) # delays for 5 seconds
driver.quit()
```

There you go. I'm not sure where I should commit that.

**Comment by jsha** [comment]
Nice! Is the script synchronous, so the page will definitely be loaded by the time we check the URL?

This should go in a new directory `test/`, with a comment at the top about how to get set up to run it. Probably run-chromium.sh should be renamed to test-chromium.sh, and do the package build, then run this script with the newly build crx. Similar to test.sh, test-chromium.sh should retain a `--justrun` flag that starts an instance with the newly built extension.

**Comment by TheNavigat** [comment]
Okay, got it.

One more thing. The new directory test/ should be created at the top directory?

**Comment by jsha** [comment]
Yep! You'll see there's already `https-everywhere-tests`, which I may move under test/ later. That directory contains a secondary Firefox extension that loads and tests the main FF one.

**Comment by TheNavigat** [comment]
I forgot to mention that yes, the script is synchronous.

**Comment by TheNavigat** [comment]
https://github.com/EFForg/https-everywhere/pull/2623

**Comment by TheNavigat** [comment]
This issue should be closed.

**Comment by semenko** [comment]
Let's keep it open for a bit -- until we get travis integration.

**Comment by TheNavigat** [comment]
Here's the Mac OS X support PR. https://github.com/EFForg/https-everywhere/pull/2722

**Comment by TheNavigat** [comment]
Regarding Travis integration, here I'm quoting @semenko comment on PR https://github.com/EFForg/https-everywhere/pull/2623

>Awesome -- thanks again!
>
>If you want to keep playing with Chrome testing, it'd be awesome to integrate this with Travis and the pre-push hook at:
>https://github.com/EFForg/https-everywhere/blob/master/test.sh
>
>You'll probably need to update the Travis apt packages to include chromium & chromium-webdriver in: https://github.com/EFForg/https-everywhere/blob/master/.travis.yml

Apart from the technical challenges, I have around a dozen logical confusions that I believe need to be resolved first.

We now have 2 different directories for testing. Firefox's testing is in ```/https-everywhere-tests/``` while Chrome's testing is in ```/test/chrome/```. Having chrome testing stuff located in ```/test/``` was suggested by @jsha

Hopping back and forth between ```/https-everywhere-tests/``` and ```/test/chrome/``` makes no sense. The logical thing is to move ```/https-everywhere-tests``` to ```/test/firefox```, however, there's still ```/https-everywhere-checker```. Maybe we should move that to ```/test/rules```?

Also, ```test.sh``` is specifically made for Firefox testing. For example, by using the ```--justrun``` argument, a Firefox instance is started. Since now we have both available (Firefox and Chromium), how should a user be able to specify which browser does he want to start?

**Comment by semenko** [comment]
If we think about this in terms of use cases, right now we have:
- People making ruleset changes
- People developing the Firefox extension itself
- People (er, mostly me) developing the Chrome extension itself
- Automated Travis testing

(Long-term, the Firefox & Chrome extensions will merge into one, but it's probably too soon to deal with that in this PR.)

I like the idea of everything in `/test/` with separate Chrome/FF directories, and a ruleset test script. Feel free to submit a PR moving those around.

We should also do more intelligent testing based on what's changed, e.g.
- Test only rules that changed (sorta done)
- Don't test FF / Chrome if there aren't corresponding changes in FF/Chrome extension code
- Make Travis test everything (FF, Chrome, Rules) always

Honestly, spinning up FF & Chrome for tiny work-in-progress changes (that I just want to push to my own testing branch) gets pretty annoying.

**Comment by TheNavigat** [comment]
Okay, let's start with moving things around.

I think we need to configure having the post-commit hook run /test/firefox/test.sh as soon as we move it. I have absolutely no idea how to configure the post-commit hook. Any advice on that?

**Comment by TheNavigat** [comment]
Here's what @jsha said on the IRC regarding the whole situation.

>jsha: I'm on vacation today and going to run to the beach soon, but my brief perspective:
>jsha: It's fine to rename test.sh to firefox-test.sh, and have a new test.sh that calls both Firefox and Chromium tests
>jsha: and firefox-test.sh should go under test/ like chromium-test.sh.
>jsha: It's just laziness on my part that I haven't organized things better so far.
>jsha: The only thing I care strongly about keeping the same is: there should be a test.sh at top level that runs all the tests; it should be possible through some command to quickly run a Firefox or Chromium profile with the latest extension installed, but I don't care what that command is

**Comment by semenko** [comment]
Sure -- let Jacob vacation :)

How about this for features:
- ./test.sh to test *everything* (with flags to select other things to test)
- ./tests/chromium-test.sh for chrome
- ./tests/firefox-test.sh for FF
- ./tests/rules-test.sh for rulesets only

Honestly, feel free to use your judgement and play with options. Clearly the testing framework has been a little neglected up until now.

**Comment by semenko** [comment]
RE pre-push hooks -- don't worry too much about it -- perhaps take a look after the rest of the testing stuff is set.

The hooks right now are all over the place (in /hooks/, but I don't think those get installed by anything) and also at https://github.com/EFForg/https-everywhere/blob/master/install-dev-dependencies.sh#L40

It might make sense to have a separate `./install-hooks.sh` file -- or Y/N prompt in `install-dev-dependencies.sh` to ask about hooks

(Since hooks allow general execution of code on a client, you have to install them yourself, otherwise cloning a repo and making a change could execute code as your user.)

**Comment by TheNavigat** [comment]
So, since I'm not familiar with Travis, does it automatically run test.sh, and if any script exits with a code other than 0, it reports that something's failed?

I'll start with moving things first, following the features you've listed, but I don't want anything to cause Travis reporting false errors because that will basically ruin everything.

**Comment by semenko** [comment]
Yep -- that's correct.

Take a look at https://github.com/EFForg/https-everywhere/blob/master/.travis.yml

And the results https://travis-ci.org/EFForg/https-everywhere

**Comment by TheNavigat** [comment]
Perfect. I'll do it right now.

I'd be grateful if you have enough time to check the PR, since there's actually a conflict hanging there...

**Comment by TheNavigat** [comment]
Here. https://github.com/EFForg/https-everywhere/pull/2731

**Comment by TheNavigat** [comment]
http://pastebin.com/dK0b6duN
http://pastebin.com/uPAC00k7

We might need to convert to docker in the near future.

**Comment by semenko** [comment]
@TheNavigat Pretty sure we're running on the Docker infra (which is what sudo: false enables).

Take a look at #2779 -- and then I think we can close this issue. Awesome work! :+1:

**Comment by TheNavigat** [comment]
Maybe we can leave it open for a bit longer, so that @jsha can let us know what other tests need to be added.

Besides, I still need to do a little bit more cleanup.

**Comment by TheNavigat** [comment]
There's one more bit of cleanup I'd like to do, move https-everywhere-checker to /test/rules. I'm not sure how to do that, because https-everywhere-checker is an independent repo. Any help, please?

**Comment by TheNavigat** [comment]
@semenko Anything?

**Comment by semenko** [comment]
Don't have time right now (I just quickly review tiny PRs when bored / waiting for science).

**Comment by TheNavigat** [comment]
Oh... Okay...

**Comment by TheNavigat** [comment]
Thanks to @mattrobenolt I made some more progress.

https://github.com/EFForg/https-everywhere/pull/2889 moves https-everywhere-checker to test/rules. It crashes because ```coverage.checker.config``` includes a wrong path for platform_certs, which is corrected by that PR. https://github.com/TheNavigat/https-everywhere-checker/pull/1

Once https-everywhere-checker's PR is merged, I need to update the commit version in ```test/rules``` repo, and then we're good to go.

On a side-note. https-everywhere-checker's branch in ```.gitmodules``` should be updates, not coverage. That's where the commit we're currently using (pre #2889 PR) exists. Not sure if I should fix this myself, maybe whoever wrote the file has a point setting it to coverage, and not updates.

**Comment by TheNavigat** [comment]
Okay so both PRs are merged, I think that's the cleanest the structure can be.

I think now's the time to wait for people to suggest more tests to add to the script.

**Comment by semenko** [comment]
Mostly fixed! Whoo!
